src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp

```diff
@@ -6278,4 +6118,185 @@ CTranslatorDXLToPlStmt::TranslateNestLoopParamList(
 	}
 	return nest_params_list;
 }
+//---------------------------------------------------------------------------
+//	@function:
+//		CTranslatorDXLToPlStmt::AcquireLocksOnChildRelations
+//
+//	@doc:
+//		Acquires locks of lockmode type either on all leaf partitions or on
+//		relations from partOids list when possible.
+//---------------------------------------------------------------------------
+void
+CTranslatorDXLToPlStmt::AcquireLocksOnChildRelations(OID oidParentRel,
+													 List *staticPartOids,
+													 LOCKMODE lockmode)
+{
+	ListCell *lc;
+
+	if (NoLock >= lockmode)
+		return;
+
+	// If static partition selection didn't take place, just acquire locks on
+	// all leaf partitions.
+	if (NIL == staticPartOids)
+	{
+		// FindAllInheritors returns a list of relation OIDs including the
+		// given rel plus all relations that inherit from it, directly or
+		// indirectly. The specified lock type is acquired on all child
+		// relations (but not on the given rel).
+		List *inhOids = gpdb::FindAllInheritors(oidParentRel, NoLock, NULL);
+
+		ForEach(lc, inhOids)
+		{
+			Oid childOid = lfirst_oid(lc);
+
+			if (childOid != oidParentRel && gpdb::IsLeafPartition(childOid))
+			{
+				gpdb::GPDBLockRelationOid(childOid, lockmode);
+			}
+		}
+		gpdb::ListFree(inhOids);
+		return;
+	}
+
+	ForEach(lc, staticPartOids)
+	{
+		OID partOid = lfirst_oid(lc);
+		gpdb::GPDBLockRelationOid(partOid, lockmode);
+	}
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CTranslatorDXLToPlStmt::EvalLockModeForChildRelations
+//
+//	@doc:
+//		Defines the lock type, which will be imposed on child partitions,
+//		based on parent's lock type.
+//
+//---------------------------------------------------------------------------
+LOCKMODE
+CTranslatorDXLToPlStmt::EvalLockModeForChildRelations(Query *parsetree,
+													  OID oidParentRel)
+{
+	ListCell *lc;
+	int rt_index;
+	LOCKMODE lockmode;
```

Comment:
Can this variable be made local only to blocks that use it?
https://github.com/arenadata/gpdb/blob/3da78c06541ca725999c8cd61eb25f732635e9c0/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L6223
https://github.com/arenadata/gpdb/blob/3da78c06541ca725999c8cd61eb25f732635e9c0/src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp#L6239